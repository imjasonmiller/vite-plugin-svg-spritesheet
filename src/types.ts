import type { ParsedPath } from "path";
import type { PluginContext } from "rollup";
import { type Config as SvgoConfig } from "svgo";

export interface Context {
  entry: string;
  plugin: PluginContext;
}

export interface ParsedSvg {
  svg: {
    "@_viewBox": string;
  };
}

export interface SvgSpriteSymbolAttrs {
  "@_id": string;
  "@_viewBox": string;
  [key: string]: string;
}

export type SvgSpritesheetMap = Map<
  string,
  { spriteId: string; spriteString: string }
>;

export interface SvgSpritesheetInclude {
  pattern: string | Array<string>;
  baseDir?: string;
}

interface SpritesheetTypesOpts {
  output: string;
  /**
   * Generate a custom declaration by passing a callback that receives a
   * `SpritesheetMap`.
   */
  generateDeclaration?: (spritesheet: SvgSpritesheetMap) => string;
}

export interface SvgSpritesheetPluginOpts {
  include: Array<SvgSpritesheetInclude>;
  output: string;
  /**
   * A custom `id` to be used as for each `<symbol />`.
   */
  customSymbolId?: (path: ParsedPath) => string;
  svgoConfig?: SvgoConfig;
  types?: SpritesheetTypesOpts;
  batchSize?: number;
}
