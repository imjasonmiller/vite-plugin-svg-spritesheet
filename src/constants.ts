export const LOG_PLUGIN_NAME = '[svg-spritesheet]'
export const DEFAULT_BATCH_SIZE = 20
export const GENERATED_FILE_HEADER = `
/**
 * WARN: Do not edit this file. It was generated from \`vite.config.ts\`.
 */
`.trim()
