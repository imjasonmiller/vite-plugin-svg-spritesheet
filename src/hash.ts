import { createHash } from 'crypto';

/**
 * Computes a SHA-256 hash of the given string content.
 *
 * @param {string} content - The input string to hash.
 * @returns {string} The hexadecimal SHA-256 hash of the input.
 */
export function computeHash(content: string): string {
  return createHash('sha256').update(content).digest('hex');
}
