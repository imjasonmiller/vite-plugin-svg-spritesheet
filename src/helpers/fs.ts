import path from 'path';
import fs from 'fs/promises';

/**
 * Checks if the given path is a valid directory.
 *
 * Safely verifies that the input is a non-empty string and that the path exists
 * and points to a directory. Returns `false` for invalid inputs or inaccessible paths.
 *
 * @param {string} path - Filesystem path to check.
 * @returns {Promise<boolean>} Resolves to `true` if the path is a directory, else `false`.
 *
 * @example
 * const isDir = await isDirectory('./logs');
 */
export async function isDirectory(path: string): Promise<boolean> {
  if (typeof path !== 'string' || path.trim() === '') {
    return false;
  }

  try {
    const stats = await fs.stat(path);
    return stats.isDirectory();
  } catch {
    return false;
  }
}

/**
 * Ensures the parent directory of the given output path exists and is writable.
 *
 * @param {string} outputPath - Full output file path whose parent directory is validated.
 * @throws {Error} If the directory does not exist or is not writable.
 * @example
 * try {
 *   await assertWritablePath(context, './public/spritesheet.svg');
 * } catch(err) {
 *   // Handle or skip file
 * }
 */
export async function assertWritablePath(outputPath: string): Promise<void> {
  const dir = path.dirname(outputPath);

  if (!(await isDirectory(dir))) {
    throw new Error(`Directory "${dir}" does not exist.`);
  }

  try {
    await fs.access(dir, fs.constants.W_OK);
  } catch {
    throw new Error(
      `Cannot write to directory "${dir}". Check your file system permissions and try again.`
    );
  }
}
