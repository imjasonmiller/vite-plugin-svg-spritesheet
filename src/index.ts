export { svgSpritesheet } from "./plugin";

export type { SvgSpritesheetPluginOpts } from "./types";
