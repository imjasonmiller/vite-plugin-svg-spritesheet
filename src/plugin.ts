import fs from 'fs/promises';
import path from 'path';
import { createHash } from 'crypto';
import { glob } from 'tinyglobby';
import picomatch from 'picomatch';
import { XMLParser, XMLBuilder } from 'fast-xml-parser';
import {
  parseSvg,
  optimizeSvg,
  cleanSymbolAttributes,
} from './helpers/parsing';
import { debounce } from './helpers/debounce';
import { writeSpritesheet } from './helpers/spritesheet';

import type { ParsedPath } from 'path';
import type { Plugin } from 'vite';
import type {
  ProcessSvgParams,
  SvgSpritesheetPluginOptions,
  SpriteMap,
  SpriteMapEntry,
  IncludeMatcher,
  SvgSpritesheetPluginContext,
} from './types';

import { LOG_PLUGIN_NAME, DEFAULT_BATCH_SIZE } from './constants';
import { normalizeError } from './helpers/error';
import { replaceColorAttributes } from './helpers/colorAttributes';

/**
 * Generate a default `id` attribute value for the `<symbol />` from path parts
 */
export function defaultSymbolId(parsedPath: ParsedPath): string {
  const dirParts = parsedPath.dir.split(path.sep).filter(Boolean);

  return ['icon', ...dirParts, parsedPath.name].join('-');
}

function isProcessingSkipped(
  context: SvgSpritesheetPluginContext,
  entry: SpriteMapEntry | undefined,
  currentHash: string,
  layerIndex: number
): boolean {
  if (!entry) {
    return false;
  }

  // Skip processing as it is overwritten by another layer
  if (entry.layerIndex > layerIndex) {
    context.logger.warn(
      `Skipping processing, as it is overwritten by another layer: ${entry.spriteId}`
    );
    return true;
  }

  // Skip processing as its contents are the same
  if (entry.hash === currentHash) {
    context.logger.warn(
      `Skipping processing, as its contents are the same: ${entry.spriteId}`
    );
    return true;
  }

  return false;
}

/**
 * Process a single SVG file: read, (optionally) optimize, parse, clean and add
 * to sprite map.
 *
 * @param params Parameters including context, file path, include info, and index
 * @returns
 */
async function processSvg({
  context,
  filePath, // rename to `entry`
  include,
  layerIndex,
}: ProcessSvgParams): Promise<void> {
  try {
    const absoluteInclude = path.resolve(include);
    const absoluteFilePath = path.resolve(absoluteInclude, filePath);
    const relativeToInclude = path.relative(absoluteInclude, absoluteFilePath);
    const parsedRelativePath = path.parse(relativeToInclude);

    const file = await fs.readFile(absoluteFilePath, 'utf8');
    // Hash file contents for comparison. MD5 is chosen for speed and
    // sufficiently low collision risk in this context. It is not intended for
    // cryptographic security.
    const hash = createHash('md5').update(file).digest('hex');
    const spriteId = context.options.customSymbolId
      ? context.options.customSymbolId(parsedRelativePath)
      : defaultSymbolId(parsedRelativePath);

    const existingEntry = context.spriteMap.get(relativeToInclude);
    if (isProcessingSkipped(context, existingEntry, hash, layerIndex)) {
      return;
    }

    const svg = optimizeSvg(context, file, filePath);
    const svgObj = parseSvg(context, svg, filePath);
    if (!svgObj) {
      context.logger.warn(
        `Parsing SVG returned empty or invalid data for file: "${filePath}". Skipping.`
      );
      return;
    }

    if (context.options.replaceColorAttributes) {
      replaceColorAttributes(svgObj);
    }

    const symbolAttrs = cleanSymbolAttributes(svgObj, spriteId);
    const spriteString = context.xmlBuilder.build({ symbol: symbolAttrs });
    context.spriteMap.set(relativeToInclude, {
      spriteId,
      spriteString,
      layerIndex,
      hash,
    });
  } catch (error) {
    const normalizedError = normalizeError(error);
    context.logger.warn(
      `Failed to process SVG file "${filePath}": ${normalizedError.message}`
    );
  }
}

const debouncedWriteSpriteSheet = debounce(
  async (args: SvgSpritesheetPluginContext) => {
    await writeSpritesheet(args);
  },
  1500,
  { leading: true, trailing: true }
);

async function handleFileEvent(
  context: SvgSpritesheetPluginContext,
  event: 'add' | 'change' | 'unlink',
  file: string,
  matchers: IncludeMatcher[]
): Promise<void> {
  for (const { matcher, include, layerIndex } of matchers) {
    if (matcher(file)) {
      switch (event) {
        case 'add':
        case 'change': {
          await processSvg({
            context,
            include,
            layerIndex,
            filePath: file,
          });
          break;
        }
        case 'unlink':
          context.spriteMap.delete(path.relative(include, file));
          break;
        default:
          context.logger.warn(
            `${LOG_PLUGIN_NAME} Unhandled event type: ${event}`
          );
          break;
      }

      debouncedWriteSpriteSheet(context);
    }
  }
}

/**
 * Vite plugin to generate SVG spritesheets from specified SVG files
 *
 * @param options - Configuration options for the plugin
 * @returns Vite plugin instance
 */
export function svgSpritesheet(options: SvgSpritesheetPluginOptions): Plugin {
  if (options.include.length === 0) {
    throw new Error('The `include` option must be a non-empty array');
  }

  const xmlOptions = {
    ignoreDeclaration: true,
    ignoreAttributes: false,
    attributeNamePrefix: '@_',
  };

  const batchSize = Math.max(1, options.batchSize ?? DEFAULT_BATCH_SIZE);

  const spriteMap: SpriteMap = new Map();
  const xmlParser = new XMLParser(xmlOptions);
  const xmlBuilder = new XMLBuilder(xmlOptions);
  const context: SvgSpritesheetPluginContext = {
    logger: console,
    options,
    spriteMap,
    xmlParser,
    xmlBuilder,
  };

  return {
    name: 'vite-plugin-svg-spritesheet',

    async buildStart() {
      context.logger = this;

      const normalizedIncludes = Array.isArray(options.include)
        ? options.include
        : [options.include];

      const resolvedInlcudes: string[] = [];

      for (const include of normalizedIncludes) {
        const resolved = path.resolve(include);

        const stats = await fs.stat(resolved).catch(() => null);
        if (!stats?.isDirectory()) {
          this.error(`The include "${include}" is not a valid directory`);
        }

        resolvedInlcudes.push(include);
      }

      for (const [layerIndex, include] of resolvedInlcudes.entries()) {
        let entries: string[] = [];
        try {
          entries = await glob('**/*.svg', {
            cwd: include,
          });
        } catch (error) {
          const normalizedError = normalizeError(error);
          this.error({
            message: `Failed to glob files for include "${include}": ${normalizedError.message}`,
          });
        }

        // Parallel processing of files in batches to avoid file exhaustion
        for (let i = 0; i < entries.length; i += batchSize) {
          const batch = entries.slice(i, i + batchSize);

          // Allow for partial failure of entries
          await Promise.allSettled(
            batch.map((entry) => {
              return processSvg({
                context,
                include,
                layerIndex,
                filePath: entry,
              });
            })
          );

          await writeSpritesheet(context);
        }
      }

      if (spriteMap.size === 0) {
        this.warn('No sprites were generated');
      } else {
        this.info(`${LOG_PLUGIN_NAME} Generated ${spriteMap.size} SVG sprites`);
      }
    },

    // NOTE: We do not use `server.watch.add` as the docs mention:
    //
    // > The Vite server watcher watches the root and skips the .git/,
    // > node_modules/, and Vite's cacheDir and build.outDir directories by
    // > default.
    //
    // Files in `src/` that match the inclusion pattern should be picked up
    //
    // See: https://vite.dev/config/server-options#server-watch
    configureServer(server) {
      // Precompile and cache all matchers
      const matchers = (
        Array.isArray(options.include) ? options.include : [options.include]
      ).map((include, index) => {
        const pattern = path.resolve(path.join(include, '**/*.svg'));

        return {
          matcher: picomatch(pattern),
          include,
          layerIndex: index,
        };
      });

      server.watcher
        .on('change', async (file) => {
          await handleFileEvent(context, 'change', file, matchers);
        })
        .on('add', async (file) => {
          await handleFileEvent(context, 'add', file, matchers);
        })
        .on('unlink', async (file) => {
          await handleFileEvent(context, 'unlink', file, matchers);
        });
    },
  };
}
