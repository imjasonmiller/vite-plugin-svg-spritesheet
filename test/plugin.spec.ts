import os from 'os'
import path from 'path'
import fs from 'fs/promises'
import { build } from 'vite'
import { describe, expect, it, afterEach, beforeEach } from 'vitest'
import { svgSpritesheet } from '../src/index'

describe('svgSpritesheet plugin (integration)', () => {
    const fixturesDir = path.resolve(__dirname, 'fixtures')

    let tempDir: string

    beforeEach(async () => {
        // Create a temporary directory that will contain the output
        tempDir = await fs.mkdtemp(
            path.join(os.tmpdir(), 'vite-plugin-svg-spritesheet-test')
        )
    })

    afterEach(async () => {
        // Remove temp directory and all contents recursively after test
        await fs.rm(tempDir, { recursive: true, force: true })
    })

    it('generates a spritesheet from fixture SVGs', async () => {
        const spritePath = path.join(tempDir, 'spritesheet.svg')

        await build({
            build: {
                write: false,
            },
            plugins: [
                svgSpritesheet({
                    include: fixturesDir,
                    output: spritePath,
                }),
            ],
        })

        const exists = await fs
            .access(spritePath)
            .then(() => true)
            .catch(() => false)

        expect(exists).toBe(true)

        const content = await fs.readFile(spritePath, 'utf8')

        // Ensure the spritesheet output doesn't unexpectedly change
        expect(content).toMatchSnapshot()

        // Check that all icon names are in the sprite
        expect(content).toContain('icon-a')
        expect(content).toContain('icon-b')
        expect(content).toContain('icon-c')
        expect(content).toContain('nested-icon-d')
    })
})
