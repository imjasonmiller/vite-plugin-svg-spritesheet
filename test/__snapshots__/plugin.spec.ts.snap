// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`svgSpritesheet plugin (integration) > generates a spritesheet from fixture SVGs 1`] = `"<svg xmlns="http://www.w3.org/2000/svg"><symbol viewBox="0 0 128 128" fill="none" id="icon-a"><path d="M78.6589 15.667C72.0289 15.667 65.5491 17.6415 60.0457 21.3387C54.5423 25.0359 50.2646 30.2883 47.7581 36.4262C45.2516 42.5642 44.6298 49.3095 45.972 55.8021C47.3142 62.2949 50.5595 68.2408 55.2943 72.8818C56.5006 74.0643 57.3275 75.5792 57.6695 77.2335C58.0114 78.8878 57.8529 80.6064 57.2143 82.1703C56.5757 83.7342 55.4858 85.0726 54.0835 86.0146C52.6813 86.9566 51.0303 87.4596 49.3411 87.4596C47.6519 87.4596 46.0009 86.9566 44.5987 86.0146C43.1964 85.0726 42.1065 83.7342 41.4679 82.1703C40.8293 80.6064 40.6708 78.8878 41.0127 77.2335C41.3547 75.5792 42.1817 74.0643 43.3881 72.8818L25.9765 55.1185C21.2418 59.7595 17.9964 65.7053 16.6542 72.198C15.312 78.6908 15.9338 85.4362 18.4403 91.5741C20.9469 97.712 25.2245 102.964 30.7279 106.662C36.2313 110.359 42.7111 112.333 49.3411 112.333C55.9711 112.333 62.4509 110.359 67.9543 106.662C73.4577 102.964 77.7355 97.712 80.242 91.5741C82.7485 85.4362 83.3703 78.6908 82.0281 72.198C80.686 65.7054 77.4406 59.7595 72.7058 55.1185C71.4995 53.936 70.6725 52.421 70.3305 50.7667C69.9885 49.1124 70.147 47.3938 70.7857 45.8299C71.4243 44.266 72.5142 42.9278 73.9164 41.9858C75.3186 41.0438 76.9696 40.5406 78.6589 40.5406C80.3481 40.5406 81.9991 41.0438 83.4013 41.9858C84.8035 42.9278 85.8934 44.266 86.5321 45.8299C87.1707 47.3938 87.3292 49.1124 86.9872 50.7667C86.6452 52.421 85.8184 53.936 84.6121 55.1185L102.024 72.8818C106.758 68.2408 110.004 62.2949 111.346 55.8021C112.688 49.3095 112.066 42.5642 109.56 36.4262C107.053 30.2883 102.775 25.0359 97.272 21.3387C91.7686 17.6415 85.2889 15.667 78.6589 15.667Z" fill="#985DF7" style="fill:#985DF7;fill:color(display-p3 0.5961 0.3647 0.9686);fill-opacity:1;"></path></symbol><symbol viewBox="0 0 128 128" fill="none" id="icon-b"><path d="M10.6543 16.9379L117.346 16.9379L64 111.062L10.6543 16.9379ZM76.4743 40.753L51.5258 40.753L64.0002 62.7631L76.4743 40.753Z" fill="#F7C558" style="fill:#F7C558;fill:color(display-p3 0.9686 0.7712 0.3451);fill-opacity:1;"></path></symbol><symbol viewBox="0 0 128 128" fill="none" id="icon-c"><path d="M64.0001 38.541C58.9648 38.541 54.0424 40.0342 49.8558 42.8316C45.6691 45.629 42.4059 49.6052 40.479 54.2571C38.5521 58.9091 38.048 64.0281 39.0303 68.9666C40.0127 73.9052 42.4374 78.4415 45.9979 82.002C49.5584 85.5625 54.0947 87.9872 59.0332 88.9695C63.9717 89.9519 69.0907 89.4478 73.7427 87.5209C78.3947 85.5939 82.3708 82.3307 85.1683 78.1441C87.9657 73.9574 89.4589 69.0352 89.4589 63.9999H77.6475V43.1044H113.553V63.9999C113.553 73.8006 110.647 83.3812 105.202 91.5301C99.757 99.6791 92.0178 106.031 82.9632 109.781C73.9085 113.532 63.945 114.513 54.3327 112.601C44.7203 110.689 35.8907 105.969 28.9606 99.0393C22.0304 92.1092 17.311 83.2796 15.399 73.6672C13.487 64.0549 14.4682 54.0913 18.2187 45.0367C21.9693 35.9821 28.3208 28.2428 36.4697 22.7978C44.6187 17.3529 54.1994 14.4468 64.0001 14.4468V38.541Z" fill="#5ECB88" style="fill:#5ECB88;fill:color(display-p3 0.3686 0.7961 0.5333);fill-opacity:1;"></path></symbol><symbol viewBox="0 0 128 128" fill="none" id="icon-d"><path d="M116.394 52.0171L96.3811 114.761H31.6189L11.6064 52.0171L64.0001 13.2393L116.394 52.0171ZM42.3624 61.952L50.8233 88.478H77.1769L85.6375 61.9519L63.9999 45.9376L42.3624 61.952Z" fill="#58BAF7" style="fill:#58BAF7;fill:color(display-p3 0.3451 0.7294 0.9686);fill-opacity:1;"></path></symbol><symbol viewBox="0 0 128 128" fill="none" id="icon-nested-d"><path d="M116.394 52.0171L96.3811 114.761H31.6189L11.6064 52.0171L64.0001 13.2393L116.394 52.0171ZM42.3624 61.952L50.8233 88.478H77.1769L85.6375 61.9519L63.9999 45.9376L42.3624 61.952Z" fill="#58BAF7" style="fill:#58BAF7;fill:color(display-p3 0.3451 0.7294 0.9686);fill-opacity:1;"></path></symbol></svg>"`;

exports[`svgSpritesheet plugin (integration) > ignores patterns listed in exclude 1`] = `"<svg xmlns="http://www.w3.org/2000/svg"><symbol viewBox="0 0 128 128" fill="none" id="icon-a"><path d="M78.6589 15.667C72.0289 15.667 65.5491 17.6415 60.0457 21.3387C54.5423 25.0359 50.2646 30.2883 47.7581 36.4262C45.2516 42.5642 44.6298 49.3095 45.972 55.8021C47.3142 62.2949 50.5595 68.2408 55.2943 72.8818C56.5006 74.0643 57.3275 75.5792 57.6695 77.2335C58.0114 78.8878 57.8529 80.6064 57.2143 82.1703C56.5757 83.7342 55.4858 85.0726 54.0835 86.0146C52.6813 86.9566 51.0303 87.4596 49.3411 87.4596C47.6519 87.4596 46.0009 86.9566 44.5987 86.0146C43.1964 85.0726 42.1065 83.7342 41.4679 82.1703C40.8293 80.6064 40.6708 78.8878 41.0127 77.2335C41.3547 75.5792 42.1817 74.0643 43.3881 72.8818L25.9765 55.1185C21.2418 59.7595 17.9964 65.7053 16.6542 72.198C15.312 78.6908 15.9338 85.4362 18.4403 91.5741C20.9469 97.712 25.2245 102.964 30.7279 106.662C36.2313 110.359 42.7111 112.333 49.3411 112.333C55.9711 112.333 62.4509 110.359 67.9543 106.662C73.4577 102.964 77.7355 97.712 80.242 91.5741C82.7485 85.4362 83.3703 78.6908 82.0281 72.198C80.686 65.7054 77.4406 59.7595 72.7058 55.1185C71.4995 53.936 70.6725 52.421 70.3305 50.7667C69.9885 49.1124 70.147 47.3938 70.7857 45.8299C71.4243 44.266 72.5142 42.9278 73.9164 41.9858C75.3186 41.0438 76.9696 40.5406 78.6589 40.5406C80.3481 40.5406 81.9991 41.0438 83.4013 41.9858C84.8035 42.9278 85.8934 44.266 86.5321 45.8299C87.1707 47.3938 87.3292 49.1124 86.9872 50.7667C86.6452 52.421 85.8184 53.936 84.6121 55.1185L102.024 72.8818C106.758 68.2408 110.004 62.2949 111.346 55.8021C112.688 49.3095 112.066 42.5642 109.56 36.4262C107.053 30.2883 102.775 25.0359 97.272 21.3387C91.7686 17.6415 85.2889 15.667 78.6589 15.667Z" fill="#985DF7" style="fill:#985DF7;fill:color(display-p3 0.5961 0.3647 0.9686);fill-opacity:1;"></path></symbol><symbol viewBox="0 0 128 128" fill="none" id="icon-b"><path d="M10.6543 16.9379L117.346 16.9379L64 111.062L10.6543 16.9379ZM76.4743 40.753L51.5258 40.753L64.0002 62.7631L76.4743 40.753Z" fill="#F7C558" style="fill:#F7C558;fill:color(display-p3 0.9686 0.7712 0.3451);fill-opacity:1;"></path></symbol><symbol viewBox="0 0 128 128" fill="none" id="icon-d"><path d="M116.394 52.0171L96.3811 114.761H31.6189L11.6064 52.0171L64.0001 13.2393L116.394 52.0171ZM42.3624 61.952L50.8233 88.478H77.1769L85.6375 61.9519L63.9999 45.9376L42.3624 61.952Z" fill="#58BAF7" style="fill:#58BAF7;fill:color(display-p3 0.3451 0.7294 0.9686);fill-opacity:1;"></path></symbol></svg>"`;
